### DISCUSSION TITLE
WebAutenticator NuGet Package Issue Resolved

### DISCUSSION THREAD

**Discussion by showtroylove** [dotMorten/WinUIEx#191] [Ideas]
I reported the NuGet package issues, but the issue was closed. The workarounds for the package were, in my estimation, more problematic than they were worth. I extended the WebAuthenticator code by implementing the Microsoft.Maui.Authentication.IWebAuthenticator interface, added a Default property like the .NET MAUI equivalent, and then used preprocessor conditions to instantiate the correct version based on the platform.

```
public sealed class WebAuthenticator : IWebAuthenticator
{
    public static WebAuthenticator Default => Instance;

    public async Task<WebAuthenticatorResult> AuthenticateAsync(WebAuthenticatorOptions o)
    {        
        if (o is null || o.Url is null || o.CallbackUrl is null)
            return new();        
        
        CancellationTokenSource cts = new(TimeSpan.FromMinutes(3)); // Optional timeout (may want to change/remove for Release build)
        var result = await AuthenticateAsync(o.Url, o.CallbackUrl, cts.Token);
        return result ?? new();
    }
}

.AddSingleton<IWebAuthenticator>(s =>
{
    return
#if WINDOWS
    MockupExpress.Services.WebAuthenticator.Default;
#else
    Microsoft.Maui.Authentication.WebAuthenticator.Default;
#endif
})

```

**Comment by dotMorten** [comment]
@showtroylove I was never able to get this to work and just tried again. You code needed some tweaks to even compile. Do you have a full sample demonstrating this in action?

**Reply by showtroylove** [reply]
Which part are you having trouble getting to work? The snippets I provided are more like instructions to follow. The bits I offered are working in my private code repository.

**Reply by dotMorten** [reply]
It never calls the winui authenticator code even though the service was registered.
When I poked around in the Maui source code I didnt see any indication that it was even possible to override the default one

**Reply by showtroylove** [reply]
I'd need to see a code snippet to provide more insight. It's not clear where you might be missing code. Dependency Injection (DI) provides the WinUI version in Windows builds. The issue I encountered was locating the WebAuth class in the Windows platform folder to avoid Visual Studio issues with references.

**Reply by dotMorten** [reply]
Here ya go: https://github.com/dotMorten/WinUIEx/compare/maui_auth_via_di
It still just crashes the same was as if I didn't register the service.

**Reply by showtroylove** [reply]
The integration approach you took is more onerous but could get the job done. You appear to be missing the DI injection of the IWebAuthenticator interface. When you inject it into the page/content classes where needed, you don’t have to worry about all the preprocessor checks #if WINDOWS {} #endif. You reduce that to one location in your IServiceCollection setup.

```csharp
public partial class MainPage : ContentPage
{
    public MainPage(IWebAuthenticator webauth)
    {
        WebAuthSvs = webauth;
        InitializeComponent();
    }

    IWebAuthenticator WebAuthSvs;  
}
```

This code becomes unnecessary:

```csharp
#if WINDOWS
    var result = await WinUIEx.WebAuthenticator.AuthenticateAsync(authorizeUri, callbackUri);
#else
    var result = await Microsoft.Maui.Authentication.WebAuthenticator.AuthenticateAsync(new WebAuthenticatorOptions()
    {
        Url = authorizeUri, 
        CallbackUrl = callbackUri
    });
#endif
```

The injection of the interface simplifies the above bit of code to this:

```csharp
var result = await WebAuthSvs.AuthenticateAsync(new WebAuthenticatorOptions()
{
    Url = authorizeUri, 
    CallbackUrl = callbackUri,
    PrefersEphemeralWebBrowserSession = DevicePlatform.iOS.ToPlatformValue(false, iOS: false)
});
```

**Reply by dotMorten** [reply]
Aaaah I see, you merely made an alternative helper method. Gotcha.

**Comment by dotMorten** [comment]
Just to address the other issue:
> The workarounds for the package were, in my estimation, more problematic than they were worth

Could you elaborate a little more what you mean by that?

> I reported the NuGet package issues, but the issue was closed.

Which issue specifically?

**Reply by showtroylove** [reply]
I initially misunderstood the NuGet package display in VS2022 but eventually got it to work using your recommendations, which led my project in an uncomfortable direction. I found a simpler solution by using just the WebAuthenticator class, which worked brilliantly and aligned with my original approach.
